### PR TITLE
encode: add some more generic impls (more tuples, references)

### PR DESCRIPTION
**Pull request by apoelstra**

### CODE REVIEW

File: src/consensus/encode.rs [hunk 1]
```diff
@@ -679,6 +679,31 @@ impl Decodable for CheckedData {
     }
 }
 
+// References
+impl<'a, T: Encodable> Encodable for &'a T {
+    fn consensus_encode<S: io::Write>(&self, s: S) -> Result<usize, Error> {
+        (&**self).consensus_encode(s)
```
Comment (sgeisler) [inline src/consensus/encode.rs:685]: I'm pretty sure the `&` isn't needed here and the method invocation magic will do the reference-taking. At least it still compiles for me on nightly and 1.29.0 without it.

**Comment by sgeisler** [comment]
Do you want to fix the breakage caused by byteorder changing its MSRV here or in a separate PR? I already have a commit that does that f06ed647c28adf872014291298d76d9f4e8d275d but didn't want to push it here. There is a high chance that  it'll need a god-mode merge for being a CI change :confused:

**Comment by apoelstra** [comment]
Thanks! I'll just cherry-pick your commit here.

It shouldn't need a god-mode merge, for what it's worth. Only enabling GA for the first time did.

**Comment by sgeisler** [comment]
> It shouldn't need a god-mode merge, for what it's worth

It really should in a secure system: I could change the CI code such that it reveals CI secrets, so it's not safe to run the PRed CI code until it is merged. But I guess it will work due to the indirection with the bash script. CI security is a mess.

**Comment by apoelstra** [comment]
I mean, there are options I can enable that'd make it impossible for outsiders to change CI. But I haven't enabled them.

File: src/consensus/encode.rs (hunk 1 above)
Comment (apoelstra) [reply src/consensus/encode.rs:685]: Yeah, it's not needed, but I think it's clearer.

**Review by sgeisler — APPROVED** [review]
ACK ef116ffbedb65dbf4cdcfd79b45a93c6ffb84cc6

Thinking about it a bit more you could probably also get away with implementing `Encodable` for `T: AsRef<E> where E: Encodable` to support all kinds of smart pointers, but the current version is much more transparent.

**Comment by apoelstra** [comment]
@sgeisler no, we specifically avoid `&[u8]`, even though it would be useful, because @TheBlueMatt complained that arrays sometimes coerced to it, causing them to be length-prefixed when arrays normally are not.

**Comment by apoelstra** [comment]
Actually, I think `T: AsRef<E>` wouldn't cover `&[u8]` ... but I'm not sure what it *does* cover, so agree that it's best to be more transparent.

**Comment by sgeisler** [comment]
It would, but I [had to look it up](https://doc.rust-lang.org/std/convert/trait.AsRef.html#impl-AsRef%3CU%3E) too :laughing:

**Comment by sanket1729** [comment]
tACK ef116ffbedb65dbf4cdcfd79b45a93c6ffb84cc6

**Review by sanket1729 — APPROVED** [review]
re tACK ef116ffbedb65dbf4cdcfd79b45a93c6ffb84cc6

**Pull request merged**